scripts/dts/gen_defines.py

```diff
@@ -969,181 +292,211 @@ def encode_zephyr_multi_level_irq(irq, irq_num):
+        if val is not None:
+            # DT_N_<node-id>_P_<prop-id>
+            macro2val[macro] = val
 
+        if prop.enum_index is not None:
+            # DT_N_<node-id>_P_<prop-id>_ENUM_IDX
+            macro2val[macro + "_ENUM_IDX"] = prop.enum_index
 
-def write_spi_dev(node):
-    # Writes SPI device GPIO chip select data if there is any
+        plen = prop_len(prop)
 
-    cs_gpio = node.spi_cs_gpio
-    if cs_gpio is None:
-        return
+        if not phandle and plen is not None:
+            # DT_N_<node-id>_P_<prop-id>_IDX_<i>
+            for i, subval in enumerate(prop.val):
+                if isinstance(subval, str):
+                    macro2val[macro + f"_IDX_{i}"] = quote_str(subval)
+                else:
+                    macro2val[macro + f"_IDX_{i}"] = subval
 
-    primary_macros = None
-    def write_spi_dev_for_ident(node, ident):
-        nonlocal primary_macros
-        if primary_macros is None:
-            primary_macros = write_controller_and_data_primary(ident,
-                                                               "CS_GPIOS", cs_gpio, "")
-        else:
-            write_controller_and_data_other(ident, "CS_GPIOS", cs_gpio, "",
-                                            primary_macros)
+        if prop.type == "phandle-array":
+            phandle_array_macros(prop, macro2val, macro)
 
-    for_each_ident(node, write_spi_dev_for_ident)
+        if plen is not None:
+            # DT_N_<node-id>_P_<prop-id>_LEN
+            macro2val[macro + "_LEN"] = plen
 
+    if macro2val:
+        out_comment("generic property macros:")
+        for macro, val in macro2val.items():
+            out_dt_define(macro, val)
+    else:
+        out_comment("(no generic property macros)")
 
-def write_clocks(node):
-    # Writes clock information.
-    #
-    # Like write_regs(), but for clocks.
-    #
-    # Most of this ought to be handled in write_props(), but the identifiers
-    # that get generated for 'clocks' are inconsistent with the with other
-    # 'phandle-array' properties.
+
+def prop2value(prop):
+    # Get the main macro value for property 'prop'
+
+    if prop.type == "string":
+        return quote_str(prop.val)
+
+    if prop.type == "int":
+        return prop.val
+
+    if prop.type == "boolean":
+        return 1 if prop.val else 0
+
+    if prop.type in ["array", "uint8-array"]:
+        return list2init(f"{val} /* {hex(val)} */" for val in prop.val)
+
+    if prop.type == "string-array":
+        return list2init(quote_str(val) for val in prop.val)
+
+    # phandle, phandles, phandle-array, path, compound: nothing
+    return None
+
+
+def prop_len(prop):
+    # Return the length of an array typed property. Return None
+    # for a scalar.
+
+    if prop.type in ["array", "uint8-array", "string-array",
+                     "phandles", "phandle-array"]:
+        return len(prop.val)
+
+    return None
+
+
+def phandle_array_macros(prop, macro2val, macro):
```

Comment:
I'd return a dictionary instead of passing in a dictionary to extend. (Could use a list of tuples too.)

Reply:
Will return a dict